run_cloud_event.go

```diff
@@ -0,0 +1,107 @@
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"strings"
+	"time"
+)
+
+type CloudEvent struct {
+	CloudEventsVersion string                 `json:"cloudEventsVersion"`
+	EventID            string                 `json:"eventID"`
+	Source             string                 `json:"source"`
+	EventType          string                 `json:"eventType"`
+	EventTypeVersion   string                 `json:"eventTypeVersion"`
+	EventTime          time.Time              `json:"eventTime"` // TODO: ensure rfc3339 format
+	SchemaURL          string                 `json:"schemaURL"`
+	ContentType        string                 `json:"contentType"`
+	Extensions         map[string]interface{} `json:"extensions"`
+	Data               interface{}            `json:"data"` // from docs: the payload is encoded into a media format which is specified by the contentType attribute (e.g. application/json)
+}
+
+func createCloudEventInput(callID, contentType, deadline string, method string, requestURL string, stdin io.Reader) (string, error) {
+	var err error
+	input := []byte("")
+	if stdin != nil {
+		input, err = ioutil.ReadAll(stdin)
+		if err != nil {
+			return "", fmt.Errorf("error reading from stdin: %v", err)
+		}
+	}
+
+	var b strings.Builder
+	enc := json.NewEncoder(&b)
+	in := &CloudEvent{
+		EventID:     callID,
+		ContentType: contentType,
+		Extensions: map[string]interface{}{
+			"protocol": CallRequestHTTP{
+				Method:     method,
+				RequestURL: requestURL,
+				Type:       "http",
+				Headers: map[string][]string{
+					"Content-Type": {contentType},
+				},
+			},
+			"deadline": deadline,
+		},
+	}
+	fmt.Println("INPUT:", string(input))
+	if len(input) == 0 {
+		// nada
+		// todo: should we leave as null, pass in empty string, omitempty or some default for the content type, eg: {} for json?
+	} else if contentType == "application/json" {
+		d := map[string]interface{}{}
+		err = json.Unmarshal(input, &d)
+		if err != nil {
+			return "", fmt.Errorf("Error unmarshalling json input: %v", err)
+		}
+		in.Data = d
+	} else {
+		in.Data = string(input)
+	}
```

Comment:
still not sure about this. from the spec when i read it this really seems like a string, where the content type defines to the user how they should decode it. the spec is not clear about this, but it does say arbitrary and not string. 

the downside being that if we do this differently than everyone else (whichever way that is) that implements the spec then users are going to run into errors parsing this field. 

i guess for every other format except json (`application/xml`, `application/vnd.hzn-3d-crossword`, `application/x-7z-compressed`, etc etc) these are going to have to be a string so that we can wedge them into json? that seems weird in itself, to me.

Reply:
https://github.com/cloudevents/spec/blob/master/http-transport-binding.md#22-data-attribute